src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts

```diff
@@ -92,25 +93,32 @@ export class DatasetService {
   }
 
   public async cacheDataset(dataset: Dataset): Promise<void> {
-    const type = this.getType(dataset.type);
-    if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
-      const spec = {
-        id: dataset.id,
-        title: dataset.title,
-        timeFieldName: dataset.timeFieldName,
-        fields: await type?.fetchFields(dataset),
-        dataSourceRef: dataset.dataSource
-          ? {
-              id: dataset.dataSource.id!,
-              name: dataset.dataSource.title,
-              type: dataset.dataSource.type,
-            }
-          : undefined,
-      } as IndexPatternSpec;
-      const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
-      if (temporaryIndexPattern) {
-        this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+    const type = this.getType(dataset?.type);
+    try {
+      if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
+        const fetchedFields = await type?.fetchFields(dataset);
+        const spec = {
+          id: dataset.id,
+          title: dataset.title,
+          timeFieldName: dataset.timeFieldName,
+          fields: fetchedFields ?? ({} as IndexPatternFieldMap),
+          areFieldsLoading: true,
+          dataSourceRef: dataset.dataSource
+            ? {
+                id: dataset.dataSource.id!,
+                name: dataset.dataSource.title,
+                type: dataset.dataSource.type,
+              }
+            : undefined,
+        } as IndexPatternSpec;
+        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
+
+        if (temporaryIndexPattern) {
+          this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+        }
       }
+    } catch (error) {
```

Comment:
Why not something like this?

```ts
public async cacheDataset(dataset: Dataset): Promise<void> {
    const type = this.getType(dataset?.type);
    try {
      const asyncType = type && type.isAsync;
      if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
        const fetchedFields = asyncType ? await type?.fetchFields(dataset) : undefined;
        const spec = {
          id: dataset.id,
          title: dataset.title,
          timeFieldName: dataset.timeFieldName,
          fields: fetchedFields ?? ({} as IndexPatternFieldMap),
          schemaLoading: asyncType ? true : false,
          dataSourceRef: dataset.dataSource
            ? {
                id: dataset.dataSource.id!,
                name: dataset.dataSource.title,
                type: dataset.dataSource.type,
              }
            : undefined,
        } as IndexPatternSpec;
        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);

        // Load schema asynchronously if it's an async index pattern
        if (asyncType && temporaryIndexPattern) {
          type.fetchFields(dataset).then((fields) => {
            temporaryIndexPattern.fields = fields;
            temporaryIndexPattern.schemaLoading = false;
            this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
          });
        }

        if (temporaryIndexPattern) {
          this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
        }
      }
    } catch (error) {
      throw new Error(`Failed to load dataset: ${dataset?.id}`);
    }
  }
  ```

This keeps the fetching of the fields colocated with the dataset service. It also avoids the situation where the loading of the field need to be coorectly specified everywhere we want to add support for async index patterns

Reply:
Nice, this is cleaner than my implementation. Earlier, I was afraid of adding lot of flags to the dataset code base. But this makes sense will update this bit.

Reply:
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)